autopush/endpoint.py

```diff
@@ -867,3 +868,67 @@ def _load_params(self):
             return json.loads(self.request.body)
         except ValueError:
             return {}
+
+
+class BridgeMessageHandler(RegistrationHandler):
+    cors_methods = "PUT"
+
+    @cyclone.web.asynchronous
+    def put(self, uaid):
+        """HTTP PUT
+
+        Process message states for a UAID
+
+        """
+        self.start_time = time.time()
+        if not self._validate_auth(uaid):
+            return self._write_unauthorized_response(
+                message="Invalid Authentication")
+        params = self._load_params()
+        if 'nak' in params:
+            d = deferToThread(self._record_naks, uaid, naks=params['nak'])
+        else:
+            d = Deferred()
+            d.callback(True)
+        d.addCallback(self._success)
+        d.addErrback(self._response_err)
+        return d
+
+    def _record_naks(self, o_uaid, naks):
+        """ Record the NAKs from the client
+
+        """
+        for nak in naks:
+            try:
+                result = self.ap_settings.fernet.decrypt(
+                    nak['id'].encode('utf8'))
+                kind, uaid, chid = result.split(":")
+                if kind != 'm':
+                    self.log.error(
+                        format="Improper messageID specified",
+                        message_id=result,
+                        uaid_hash=hasher(uaid),
+                        channel_id=chid,
+                        message=json.dumps(nak))
+                    continue
+                try:
+                    self.log.info(
+                        format="NAK",
+                        uaid_hash=hasher(uaid),
+                        channel_id=chid,
+                        user_agent=self.request.user_agent,
+                        message_id=result,
+                        message=nak.get("msg", None),
+                        code=nak.get("code", 0))
```

Comment:
I can see where a NAK may contain just the channelid. (e.g. the client either cannot determine the code or msg, or fails to properly send that data). This could still be useful for flow debugging.